File: composeApp/src/androidMain/kotlin/com/selenite/scanner/data/ContentSignature.kt
package com.selenite.scanner.data

import android.util.Log
import java.io.File
import java.io.InputStream
import java.io.RandomAccessFile
import java.security.KeyFactory
import java.security.MessageDigest
import java.security.Signature
import java.security.spec.X509EncodedKeySpec
import java.util.Base64

/**
 * Verifies that a content release was produced by Miror's release pipeline.
 *
 * Miror Link relays content packages phone to phone over an untrusted radio, and the
 * digests carried in the wire protocol are supplied by the sender -- they prove the bytes
 * arrived intact, never that the sender is entitled to author them. Without this check a
 * modified client could hand a peer any three files that merely *look* structurally
 * valid, and they would be merged into the live catalog permanently.
 *
 * Verification runs **before** `cards.db` is handed to SQLite. Hashing a file is only
 * reading bytes -- nothing is parsed, decoded, or opened as a database -- so a package
 * that fails this check never reaches a parser that could be attacked by its contents.
 *
 * ECDSA P-256 rather than Ed25519 because `minSdk` is 26 and Ed25519 only reached the
 * platform JCA in API 33. `SHA256withECDSA` has been present since API 1.
 *
 * The signed message must match `scripts/ContentUpdate/content_signature.py` byte for
 * byte. Nothing in either codebase would fail if the two drifted -- releases would simply
 * stop verifying on every device at once, after shipping. The only thing that turns that
 * into a test failure first is the pinned fixture in
 * `androidUnitTest/.../data/ContentSignatureTest.kt` (`Fixture`, exercised by
 * `aPackageSignedByTheReleasePipelineVerifies`), whose signature line was produced by the
 * Python signer and is checked here by the Kotlin verifier.
 */
object ContentSignature {

    private const val TAG = "ContentSignature"

    /**
     * Domain separator: a signature made for anything else must not verify here.
     *
     * The trailing NUL terminates the label so no future prefix can collide with it.
     * It is appended rather than written inline: a raw NUL inside a source literal
     * makes the file binary to git and grep. Must match DOMAIN in content_signature.py.
     */
    private val DOMAIN =
        "miror.content.v1".toByteArray(Charsets.US_ASCII) + byteArrayOf(0)

    private const val SIGNATURE_TAG = "#miror-sig-v1"

    private val NEWLINE = '\n'.code.toByte()

    /**
     * How far back from the end of the ids file the signature line may start.
     *
     * A real line is a tag, a short key id and a base64 P-256 signature -- under 130
     * bytes. The window is generous enough to be future-proof and small enough that a
     * hostile file cannot make this allocation matter.
     */
    private const val MAX_SIGNATURE_TAIL_BYTES = 4096
    private const val MAX_SIGNATURE_LINE_BYTES = 512

    /** `#miror-sig-`, the version-agnostic prefix, as bytes for the streaming scan. */
    private val SIGNATURE_MARKER = "#miror-sig-".toByteArray(Charsets.US_ASCII)
    private const val ALGORITHM = "SHA256withECDSA"
    private const val KEY_ALGORITHM = "EC"

    /**
     * Accepted verification keys, by id.
     *
     * These are public keys; they ship in the APK, are extractable from it, and are
     * meant to be public. Private signing material is not part of the source-available
     * publication.
     *
     * More than one key is accepted so a key can be rotated: a new key must ship here
     * **before** the pipeline starts signing with it, or receivers on older builds would
     * silently stop accepting content. `e37bb53f` is that reserve and has never signed
     * anything.
     */
    private val ACCEPTED_KEYS: Map<String, String> = mapOf(
        "0140a43a" to
            "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEOocivjh2kxNix2msUUa+WMy4jlK5" +
            "kh3h/hSF+pgTE6lWhfZ4ezHTta8Cb/c+6V4pG2ASXfC4B/83X+ZgE6dGUA==",
        "e37bb53f" to
            "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE+Sp6mqrcInb34ABINbMCwrVx+94w" +
            "nFZYWtUI6uijOjFKJbtzfLCHf+FGV0evv3maf3ETi/youJ45KkkNA2Mpxw==",
    )

    /**
     * True when [dir] holds a complete release whose signature covers exactly these
     * bytes at exactly [expectedVersion].
     *
     * Any failure -- missing file, absent signature, unknown key, tampered byte, wrong
     * version -- returns false. There is deliberately no "unsigned is acceptable" branch:
     * no unsigned package has ever shipped, so allowing one would create a permanently
     * weaker path for no compatibility benefit.
     */
    fun verify(dir: File, expectedVersion: Int): Boolean {
        return try {
            verifyOrThrow(dir, expectedVersion)
            true
        } catch (failure: Throwable) {
            Log.w(TAG, "Content signature rejected: ${failure.message}")
            false
        }
    }

    private fun verifyOrThrow(dir: File, expectedVersion: Int) {
        require(expectedVersion > 0) { "content version must be positive" }
        ContentAssets.REQUIRED_FILES.forEach { name ->
            require(File(dir, name).isFile) { "release is missing $name" }
        }

        val idsFile = File(dir, ContentAssets.INDEX_IDS)
        val located = locateSignature(idsFile)
        val parsed = parseSignatureLine(located.line)
        val encodedKey = ACCEPTED_KEYS[parsed.keyId]
            ?: error("signature uses unknown key id ${parsed.keyId}")

        val message = buildMessage(
            version = expectedVersion,
            cardsDigest = digestOf(File(dir, ContentAssets.CARDS_DB)),
            indexDigest = digestOf(File(dir, ContentAssets.INDEX_BIN)),
            idsDigest = digestOfPrefix(idsFile, located.bodyLength),
        )

        val publicKey = KeyFactory.getInstance(KEY_ALGORITHM).generatePublic(
            X509EncodedKeySpec(Base64.getDecoder().decode(encodedKey)),
        )
        val verified = Signature.getInstance(ALGORITHM).run {
            initVerify(publicKey)
            update(message)
            verify(parsed.signature)
        }
        if (!verified) error("signature does not match these files at v$expectedVersion")
        Log.i(TAG, "Content signature verified for v$expectedVersion with key ${parsed.keyId}")
    }

    /** Where the signed body ends, and the signature line that follows it. */
    internal class LocatedSignature(val bodyLength: Long, val line: String)

    /**
     * Finds the trailing signature by reading only the tail of the file.
     *
     * Deliberately never loads the ids file whole. The protocol permits a content file
     * of up to [MirorLinkProtocol.MAX_CONTENT_FILE_BYTES], and a peer whose package the
     * user has consented to receive controls every byte of it. Reading it into a single
     * array -- then decoding it to UTF-16, splitting it, and Base64-decoding a chunk of
     * it -- turns a bounded disk cost into a predictable heap exhaustion, which on a
     * Pixel XL-class device is a crash rather than a rejection.
     *
     * Consent covers spending disk on a package. It does not cover being killed by it.
     */
    internal fun locateSignature(file: File): LocatedSignature {
        val length = file.length()
        require(length > 0) { "gallery ids are empty" }
        val tailSize = minOf(length, MAX_SIGNATURE_TAIL_BYTES.toLong()).toInt()
        val tail = ByteArray(tailSize)
        RandomAccessFile(file, "r").use { handle ->
            handle.seek(length - tailSize)
            handle.readFully(tail)
        }
        require(tail[tail.size - 1] == NEWLINE) { "gallery ids must end with a newline" }

        // Index of the newline that terminates the line before the last one.
        var separator = -1
        for (index in tail.size - 2 downTo 0) {
            if (tail[index] == NEWLINE) {
                separator = index
                break
            }
        }
        if (separator < 0) {
            // The final line is longer than the tail we are willing to read, so it cannot
            // be a signature line -- a real one is well under a hundred bytes.
            error("no signature line within the last $MAX_SIGNATURE_TAIL_BYTES bytes")
        }
        val lineBytes = tail.copyOfRange(separator + 1, tail.size - 1)
        require(lineBytes.size <= MAX_SIGNATURE_LINE_BYTES) { "signature line is too long" }
        val line = String(lineBytes, Charsets.UTF_8).trim()
        require(line.startsWith(SIGNATURE_TAG)) { "release carries no signature line" }

        // Body length in absolute file terms: everything up to and including the newline
        // that precedes the signature line.
        return LocatedSignature(
            bodyLength = length - tailSize + separator + 1,
            line = line,
        )
    }

    /**
     * The exact bytes that are signed.
     *
     * All three digests live in one message on purpose. Per-file signatures would allow
     * mix-and-match -- one release's genuine `cards.db` beside another's genuine gallery
     * and genuine signature, every part authentic -- which is precisely the torn state
     * that yields cards the catalog lists but the scanner cannot find.
     *
     * The version is inside the message for the same reason: it stops a peer relabelling
     * a genuine old package as a far newer one, which would otherwise pin the device off
     * updates permanently.
     */
    internal fun buildMessage(
        version: Int,
        cardsDigest: ByteArray,
        indexDigest: ByteArray,
        idsDigest: ByteArray,
    ): ByteArray {
        require(cardsDigest.size == 32 && indexDigest.size == 32 && idsDigest.size == 32) {
            "content digests must be 32 bytes"
        }
        val message = ByteArray(DOMAIN.size + 4 + 96)
        var offset = 0
        DOMAIN.copyInto(message, offset); offset += DOMAIN.size
        // Big-endian, matching struct.pack(">I", version) on the Python side.
        message[offset++] = (version ushr 24).toByte()
        message[offset++] = (version ushr 16).toByte()
        message[offset++] = (version ushr 8).toByte()
        message[offset++] = version.toByte()
        cardsDigest.copyInto(message, offset); offset += 32
        indexDigest.copyInto(message, offset); offset += 32
        idsDigest.copyInto(message, offset)
        return message
    }

    internal class SplitIds(val body: ByteArray, val signatureLine: String?)

    /**
     * Splits the ids file into the signed body and the trailing signature line.
     *
     * Only the final line counts. Position is part of the contract, for two reasons: the
     * hashed input has to be canonical, and the signature must never displace a gallery
     * row -- `printlens_v3_families.tsv` addresses the gallery by row position, so a
     * prepended line would make PrintLens rerank against the wrong card with no error
     * anywhere.
     */
    internal fun splitSignature(raw: ByteArray): SplitIds {
        require(raw.isNotEmpty() && raw.last() == '\n'.code.toByte()) {
            "gallery ids must end with a newline"
        }
        var lastLineStart = 0
        for (index in raw.size - 2 downTo 0) {
            if (raw[index] == '\n'.code.toByte()) {
                lastLineStart = index + 1
                break
            }
        }
        val lastLine = String(raw, lastLineStart, raw.size - 1 - lastLineStart, Charsets.UTF_8).trim()
        return if (lastLine.startsWith(SIGNATURE_TAG)) {
            SplitIds(raw.copyOfRange(0, lastLineStart), lastLine)
        } else {
            SplitIds(raw, null)
        }
    }

    private class ParsedSignature(val keyId: String, val signature: ByteArray)

    private fun parseSignatureLine(line: String): ParsedSignature {
        val parts = line.trim().split(Regex("\\s+"))
        require(parts.size == 3) { "malformed signature line" }
        require(parts[0] == SIGNATURE_TAG) { "unsupported signature tag ${parts[0]}" }
        val decoded = runCatching { Base64.getDecoder().decode(parts[2]) }
            .getOrElse { error("signature is not valid base64") }
        require(decoded.isNotEmpty()) { "signature is empty" }
        return ParsedSignature(parts[1], decoded)
    }

    /**
     * Streams the first [byteCount] bytes through SHA-256.
     *
     * Used for the ids body so the signed prefix is hashed without ever materialising it.
     */
    private fun digestOfPrefix(file: File, byteCount: Long): ByteArray {
        require(byteCount >= 0 && byteCount <= file.length()) { "ids body length is out of range" }
        val digest = MessageDigest.getInstance("SHA-256")
        val marker = SIGNATURE_MARKER
        // Rolling match against the signature marker at a line start, so a second marker
        // is caught without holding the body in memory. "The last line is the signature"
        // and "there is one signature" are different claims; tools that assume the second
        // would disagree about which line is authoritative.
        var atLineStart = true
        var matched = 0
        file.inputStream().use { stream ->
            val buffer = ByteArray(64 * 1024)
            var remaining = byteCount
            while (remaining > 0) {
                val wanted = minOf(buffer.size.toLong(), remaining).toInt()
                val read = stream.read(buffer, 0, wanted)
                if (read <= 0) error("ids file ended before its signed body did")
                digest.update(buffer, 0, read)
                for (index in 0 until read) {
                    val byte = buffer[index]
                    if (byte == NEWLINE) {
                        atLineStart = true
                        matched = 0
                        continue
                    }
                    if (atLineStart) {
                        if (byte == marker[matched]) {
                            matched++
                            if (matched == marker.size) {
                                error("gallery ids carry a second signature marker")
                            }
                        } else {
                            atLineStart = false
                            matched = 0
                        }
                    }
                }
                remaining -= read
            }
        }
        return digest.digest()
    }

    /** Streams the file through SHA-256. Reads bytes only -- never parses or opens. */
    private fun digestOf(file: File): ByteArray {
        val digest = MessageDigest.getInstance("SHA-256")
        file.inputStream().use { stream: InputStream ->
            val buffer = ByteArray(1024 * 1024)
            while (true) {
                val read = stream.read(buffer)
                if (read <= 0) break
                digest.update(buffer, 0, read)
            }
        }
        return digest.digest()
    }
}

File: scripts/ContentUpdate/content_signature.py
"""
Detached signatures for Miror content releases.

A content release is three files that must move together: ``cards.db``,
``master_index_v3.bin`` and ``master_index_v3_ids.txt``. Miror Link relays those files
between phones over an untrusted radio, so the receiver cannot trust the sender to tell
the truth about them. The signature is what lets a device check, offline and unaided,
that a package came from this pipeline regardless of which peer handed it over.

Design notes that are load-bearing rather than stylistic:

*One signature over all three digests.*  Per-file signatures would allow mix-and-match --
v12's genuine ``cards.db`` beside v13's genuine ``.bin`` and v13's genuine signature,
every part authentic. Binding all three into a single signed message makes that
combination unsigned. This is exactly the torn state ``ContentUpdateManager`` warns about
(a catalog naming cards the gallery cannot scan), now structurally impossible.

*The version is inside the message.*  Otherwise a peer can relabel a genuine v12 package
as v2147483647, and because every version check in the app is relative ("newer than
mine") and permanent, that would pin the device off updates forever.

*The signature lives in the package, appended last.*  It has to travel inside the files so
a device can relay content it received without being able to forge it -- a signature
carried in the wire protocol would die after one relay hop. It goes in the ids file
because that is the only one of the three that can be read safely before verification:
``cards.db`` would mean opening SQLite (the parse being gated) and the ``.bin`` length is
checked exactly against the row count. And it is *appended*, never prepended, because
``printlens_v3_families.tsv`` addresses the global gallery by row position -- shifting
those positions would make PrintLens rerank against the wrong card, silently.

ECDSA P-256 rather than Ed25519: Miror's ``minSdk`` is 26, and Ed25519 only reached the
Android platform JCA in API 33. ``SHA256withECDSA`` has been available since API 1.
"""

from __future__ import annotations

import base64
import hashlib
import struct
from pathlib import Path

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec

CARDS_DB = "cards.db"
INDEX_BIN = "master_index_v3.bin"
INDEX_IDS = "master_index_v3_ids.txt"
RELEASE_FILES = (CARDS_DB, INDEX_BIN, INDEX_IDS)

SIGNATURE_PREFIX = "#miror-sig-"
SIGNATURE_TAG = "miror-sig-v1"

# Domain separation: a signature produced for any other purpose must never be
# reinterpretable as a content signature.
DOMAIN = b"miror.content.v1\x00"


class ContentSignatureError(Exception):
    """Raised for any malformed, missing or invalid release signature."""


# --------------------------------------------------------------------------- ids parsing


def is_id_line(line: str) -> bool:
    """True for a line denoting an actual gallery row (not blank, not metadata)."""
    stripped = line.strip()
    return bool(stripped) and not stripped.startswith("#")


def parse_ids(text: str) -> list[str]:
    """Gallery IDs in file order, with blank lines and metadata removed.

    Mirrors ``GalleryIds.parse`` on the Kotlin side. The two must agree exactly: the row
    count derived here sizes the ``.bin`` byte-count assertion, and the row count derived
    there sizes the runtime fp16 gallery.
    """
    return [line.strip() for line in text.splitlines() if is_id_line(line)]


def split_signature(raw: bytes) -> tuple[bytes, str | None]:
    """Split ids-file bytes into (body, signature_line).

    ``body`` is everything up to and including the newline that precedes the signature
    line, and is exactly what gets hashed. Only the final line is considered, so a
    signature placed anywhere else is not a signature -- position is part of the contract,
    both to keep the hashed input canonical and to guarantee gallery row positions never
    shift.
    """
    if not raw.endswith(b"\n"):
        raise ContentSignatureError("gallery ids must end with a newline")
    without_final_newline = raw[:-1]
    last_line_start = without_final_newline.rfind(b"\n") + 1
    try:
        last_line = without_final_newline[last_line_start:].decode("utf-8")
    except UnicodeDecodeError as exc:
        raise ContentSignatureError("gallery ids are not valid UTF-8") from exc

    if last_line.strip().startswith(SIGNATURE_PREFIX):
        body = raw[:last_line_start]
        assert_no_signature_in_body(body)
        return body, last_line.strip()
    assert_no_signature_in_body(raw)
    return raw, None


def assert_no_signature_in_body(body: bytes) -> None:
    """A signed release carries exactly one signature line, and it is the last one.

    A second marker is not forgeable -- the body is covered by the signature, so any
    marker inside it was authored by this pipeline. It is still refused, because "the
    final line is the signature" and "there is one signature" are different claims, and
    tools that reasonably assume the second would disagree about which line is
    authoritative. Canonical inputs are what make signature schemes debuggable.
    """
    for number, line in enumerate(body.split(b"\n"), start=1):
        if line.strip().startswith(SIGNATURE_PREFIX.encode("ascii")):
            raise ContentSignatureError(
                f"gallery ids carry a second signature marker at line {number}"
            )


# ------------------------------------------------------------------------------ hashing


def sha256_file(path: Path) -> bytes:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for block in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(block)
    return digest.digest()


def ids_body_bytes(path: Path) -> bytes:
    """The gallery rows of an ids file, without any trailing signature line."""
    body, _ = split_signature(Path(path).read_bytes())
    return body


def sha256_ids_body_hex(path: Path) -> str:
    """Hex digest of the ids **body** -- the value recorded and compared everywhere.

    Not the whole-file digest, and the distinction is load-bearing. ``cards.db`` carries
    ``index_ids_sha256`` in its ``content_meta``, and it has to be stamped *before* the
    release is signed, because the database's own digest is one of the three inside the
    signature. Stamping afterwards would invalidate the thing being stamped.

    So the only value that can be both recorded at build time and re-derived at verify
    time is the body digest. It is also the semantically right one: this hash identifies
    the gallery content, while the signature covers the envelope around it.

    For an unsigned file this equals the whole-file digest, so existing baselines and
    manifests keep verifying unchanged.
    """
    return hashlib.sha256(ids_body_bytes(path)).hexdigest()


def build_message(version: int, db_digest: bytes, bin_digest: bytes, ids_digest: bytes) -> bytes:
    """The exact byte string that gets signed and verified.

    Any change here is a wire-format break and must be matched byte-for-byte by
    ``ContentSignature.kt``. The round-trip test exists to catch drift.
    """
    if version <= 0:
        raise ContentSignatureError(f"content version must be positive, got {version}")
    for name, digest in (("cards.db", db_digest), ("index bin", bin_digest), ("ids", ids_digest)):
        if len(digest) != 32:
            raise ContentSignatureError(f"{name} digest must be 32 bytes, got {len(digest)}")
    return DOMAIN + struct.pack(">I", version) + db_digest + bin_digest + ids_digest


def release_message(release_dir: Path, version: int) -> bytes:
    """Builds the signed message for a release directory, ignoring any existing signature."""
    directory = Path(release_dir)
    for name in RELEASE_FILES:
        if not (directory / name).is_file():
            raise ContentSignatureError(f"release is missing {name}")
    body, _ = split_signature((directory / INDEX_IDS).read_bytes())
    return build_message(
        version,
        sha256_file(directory / CARDS_DB),
        sha256_file(directory / INDEX_BIN),
        hashlib.sha256(body).digest(),
    )


# ------------------------------------------------------------------------------- keys


def load_private_key(pem: bytes | str) -> ec.EllipticCurvePrivateKey:
    data = pem.encode("utf-8") if isinstance(pem, str) else pem
    key = serialization.load_pem_private_key(data, password=None)
    if not isinstance(key, ec.EllipticCurvePrivateKey):
        raise ContentSignatureError("signing key is not an elliptic-curve private key")
    if not isinstance(key.curve, ec.SECP256R1):
        raise ContentSignatureError(f"signing key must be P-256, got {key.curve.name}")
    return key


def load_public_key(pem: bytes | str) -> ec.EllipticCurvePublicKey:
    data = pem.encode("utf-8") if isinstance(pem, str) else pem
    key = serialization.load_pem_public_key(data)
    if not isinstance(key, ec.EllipticCurvePublicKey):
        raise ContentSignatureError("verification key is not an elliptic-curve public key")
    return key


def public_key_der_base64(key: ec.EllipticCurvePublicKey) -> str:
    """X.509 SubjectPublicKeyInfo, base64. This is the form the app embeds."""
    return base64.b64encode(
        key.public_bytes(
            encoding=serialization.Encoding.DER,
            format=serialization.PublicFormat.SubjectPublicKeyInfo,
        )
    ).decode("ascii")


def generate_keypair() -> tuple[str, str]:
    """Returns (private_pem, public_pem). Used once; the private key is never regenerated."""
    private = ec.generate_private_key(ec.SECP256R1())
    private_pem = private.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption(),
    ).decode("ascii")
    public_pem = private.public_key().public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    ).decode("ascii")
    return private_pem, public_pem


# -------------------------------------------------------------------- sign and verify


def format_signature_line(key_id: str, signature: bytes) -> str:
    if not key_id or any(character.isspace() for character in key_id):
        raise ContentSignatureError("key id must be non-empty and contain no whitespace")
    return f"#{SIGNATURE_TAG} {key_id} {base64.b64encode(signature).decode('ascii')}"


def parse_signature_line(line: str) -> tuple[str, bytes]:
    parts = line.strip().split()
    if len(parts) != 3:
        raise ContentSignatureError(f"malformed signature line: {line!r}")
    tag, key_id, encoded = parts
    if tag != f"#{SIGNATURE_TAG}":
        raise ContentSignatureError(f"unsupported signature tag: {tag}")
    try:
        signature = base64.b64decode(encoded, validate=True)
    except Exception as exc:  # noqa: BLE001 - any decode failure is the same outcome
        raise ContentSignatureError("signature is not valid base64") from exc
    return key_id, signature


def sign_release(release_dir: Path, version: int, private_pem: bytes | str, key_id: str) -> str:
    """Signs a release in place, appending the signature line to the ids file.

    Re-signing an already-signed release replaces the existing line rather than stacking
    a second one, so this is safe to re-run.
    """
    directory = Path(release_dir)
    ids_path = directory / INDEX_IDS
    body, _ = split_signature(ids_path.read_bytes())

    message = build_message(
        version,
        sha256_file(directory / CARDS_DB),
        sha256_file(directory / INDEX_BIN),
        hashlib.sha256(body).digest(),
    )
    signature = load_private_key(private_pem).sign(message, ec.ECDSA(hashes.SHA256()))
    line = format_signature_line(key_id, signature)

    # Written body-first so a crash mid-write cannot leave a file whose signature does
    # not describe the bytes above it.
    ids_path.write_bytes(body + line.encode("ascii") + b"\n")
    return line


def verify_release(release_dir: Path, version: int, public_keys: dict[str, str]) -> str:
    """Verifies a signed release. Returns the key id used, or raises.

    ``public_keys`` maps key id to PEM. Multiple entries are supported so a key can be
    rotated: ship the new key in the app first, then start signing with it.
    """
    directory = Path(release_dir)
    for name in RELEASE_FILES:
        if not (directory / name).is_file():
            raise ContentSignatureError(f"release is missing {name}")

    body, signature_line = split_signature((directory / INDEX_IDS).read_bytes())
    if signature_line is None:
        raise ContentSignatureError("release carries no signature line")

    key_id, signature = parse_signature_line(signature_line)
    pem = public_keys.get(key_id)
    if pem is None:
        raise ContentSignatureError(f"signature uses unknown key id {key_id!r}")

    message = build_message(
        version,
        sha256_file(directory / CARDS_DB),
        sha256_file(directory / INDEX_BIN),
        hashlib.sha256(body).digest(),
    )
    try:
        load_public_key(pem).verify(signature, message, ec.ECDSA(hashes.SHA256()))
    except InvalidSignature as exc:
        raise ContentSignatureError(
            "signature does not match these files at this version"
        ) from exc
    return key_id
